### PR TITLE
Fix mergeIterator.Seek()

### PR DESCRIPTION
**Pull request by codesome**
#### What this PR does

mergeIterator.Seek() was skipping samples in some cases. This was happening because in the Seek() method we were discarding the current batch from mergeIterator if the batch's first sample was after the seek time.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [X] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by charleskorn — COMMENTED** [review]
Thanks for tracking this down and fixing it!

Looks like `chunkIterator.Seek` has a similar check (and therefore I assume the same bug) - should we change that as well?

I'm also wondering if we should clarify the description of `iterator.Seek` in `batch.go` to be clear that it'll seek to the batch with a sample at or after the desired timestamp, not just a batch at or after the desired timestamp:

```diff
type iterator interface {
-	// Seek to the batch at (or after) time t.
+	// Seek to the batch with sample at (or after) time t.
	Seek(t int64, size int) chunkenc.ValueType
```

File: pkg/querier/batch/merge_test.go [hunk 1]
```diff
@@ -61,3 +63,47 @@ func TestMergeHarder(t *testing.T) {
 		})
 	}
 }
+
+// TestBugInMergeIterator tests a bug while calling Seek() on mergeIterator.
+func TestBugInMergeIteratorSeek(t *testing.T) {
```
Comment (dimitarvdimitrov) [inline pkg/querier/batch/merge_test.go:68]: nit on naming: aren't all tests testing for a bug? 
```suggestion
// TestMergeIterator tests a bug while calling Seek() on mergeIterator.
func TestMergeIteratorSeek(t *testing.T) {
```

File: pkg/querier/batch/merge_test.go [hunk 2]
```diff
@@ -61,3 +63,47 @@ func TestMergeHarder(t *testing.T) {
...
+			require.NoError(t, err)
+			require.NoError(t, chk.UnmarshalFromBuf(ch.Bytes()))
+			return chk.NewIterator(reuse)
+		}))
+	}
+
+	c3It := NewGenericChunkMergeIterator(nil, labels.EmptyLabels(), genericChunks)
+
+	c3It.Seek(15)
+	// These Next() calls are necessary to reproduce the bug.
```
Comment (dimitarvdimitrov) [inline pkg/querier/batch/merge_test.go:96]: are they necessary so that we cross the chunk boundary or?

**Review by dimitarvdimitrov — COMMENTED** [review]
thanks!

File: pkg/querier/batch/merge_test.go (hunk 2 above)
Comment (codesome) [reply pkg/querier/batch/merge_test.go:96]: Yes I think so. These three Next() cross from chunk 1 to 2. I will probably simply the test if it still reproduces the earlier bug

Comment (codesome) [reply pkg/querier/batch/merge_test.go:96]: Looks like there is something specific with these three Next() that reproduces it with this input. I don't think I want to dig further into details of why at the moment :)

**Comment by codesome** [comment]
> Looks like chunkIterator.Seek has a similar check (and therefore I assume the same bug) - should we change that as well?

I updated it similar to mergeIterator, but I am not sure if we can reproduce the bug in that one with same input since chunkIterator is just for a single chunk. I think we can leave it as it is with the potential fix. Not sure if we want to invest time in trying to break chunkIterator.Seek.

**Review by charleskorn — APPROVED** [review]

**Comment by charleskorn** [comment]
>I think we can leave it as it is with the potential fix. Not sure if we want to invest time in trying to break chunkIterator.Seek.

I think it'd be good to add a test to ensure we don't reintroduce this bug in the future, but this can come in a follow-up PR.

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM. Plus one on the test for chunkIterator (but better to merge this now before the weekend)

Maybe we can create some general test for all iterators which ensures that they comply with the interface. I think I found a bug in memoizingIterator in that it skips the first sample if you first call Next() on it. But we never use that, so it's still not causing problems. I can have a go at that next week.

**Comment by codesome** [comment]
> but this can come in a follow-up PR

+1

**Comment by charleskorn** [comment]
> > but this can come in a follow-up PR
> 
> +1

I've added a test in https://github.com/grafana/mimir/pull/9040

**Pull request merged**